lib/rubygems/user_interaction.rb

```diff
@@ -575,6 +575,57 @@ def done
   end
 
   ##
+  # A progress reporter that behaves nicely with threaded downloading.
+
+  class ThreadedDownloadReporter
```

Comment:
I explained how I'm planning to resolve this in [a comment on the PR itself](https://github.com/rubygems/rubygems/pull/1898#issuecomment-307607207). (Linking here so it's easier for people to find.)